File: src/superbid_collector/identity_hints.py
from __future__ import annotations

import re
from dataclasses import dataclass

IDENTITY_HINT_GUARDRAIL = "AUTOMATED_IDENTITY_HINT_NOT_HUMAN_EVIDENCE_OR_MATCH"

ENGINE_CC_RE = re.compile(r"\b(\d{3,5})\s*CC\b", re.I)

_MANUAL_TRANSMISSION_RE = re.compile(r"\b(?:MT|MANUAL|MEC[ÁA]NIC[AO])\b", re.I)
_AUTOMATIC_TRANSMISSION_RE = re.compile(
    r"\b(?:AT|TP|CVT|DCT|DSG|AUT|AUTOM[ÁA]TIC[AO])\b", re.I
)
_4X4_RE = re.compile(r"\b(?:4\s*[Xx]\s*4|4WD|AWD)\b", re.I)
_4X2_RE = re.compile(r"\b(?:4\s*[Xx]\s*2|2WD)\b", re.I)
_HYBRID_RE = re.compile(r"\b(?:H[IÍ]BRID[AO]|HYBRID|HEV|PHEV)\b", re.I)
_ELECTRIC_RE = re.compile(r"\b(?:EL[EÉ]CTRIC[AO]|ELECTRIC|EV|BEV)\b", re.I)
_DIESEL_RE = re.compile(r"\b(?:DI[EÉ]SEL|DIESEL)\b", re.I)
_CNG_RE = re.compile(r"\b(?:GNV|GNC|CNG|GAS\s+NATURAL)\b", re.I)
_GASOLINE_RE = re.compile(r"\b(?:GASOLINA|GASOLINE|PETROL)\b", re.I)


@dataclass(frozen=True)
class VehicleIdentityHints:
    engine_cc: int | None = None
    transmission: str | None = None
    drivetrain: str | None = None
    fuel: str | None = None

    def as_dict(self) -> dict[str, int | str | None]:
        return {
            "engine_cc": self.engine_cc,
            "transmission": self.transmission,
            "drivetrain": self.drivetrain,
            "fuel": self.fuel,
        }


def _unique_engine_cc(text: str) -> int | None:
    values = {int(match.group(1)) for match in ENGINE_CC_RE.finditer(text)}
    return next(iter(values)) if len(values) == 1 else None


def _exclusive_flag(text: str, rules: tuple[tuple[str, re.Pattern[str]], ...]) -> str | None:
    values = {label for label, pattern in rules if pattern.search(text)}
    return next(iter(values)) if len(values) == 1 else None


def _fuel_hint(text: str) -> str | None:
    # Hybrid/electric labels are treated as higher-order propulsion declarations.
    if _HYBRID_RE.search(text):
        return "HYBRID"
    if _ELECTRIC_RE.search(text):
        return "ELECTRIC"
    return _exclusive_flag(
        text,
        (
            ("DIESEL", _DIESEL_RE),
            ("CNG", _CNG_RE),
            ("GASOLINE", _GASOLINE_RE),
        ),
    )


def extract_vehicle_identity_hints(text: str | None) -> VehicleIdentityHints:
    raw = str(text or "")
    return VehicleIdentityHints(
        engine_cc=_unique_engine_cc(raw),
        transmission=_exclusive_flag(
            raw,
            (
                ("MANUAL", _MANUAL_TRANSMISSION_RE),
                ("AUTOMATIC", _AUTOMATIC_TRANSMISSION_RE),
            ),
        ),
        drivetrain=_exclusive_flag(raw, (("4X4_AWD", _4X4_RE), ("4X2_2WD", _4X2_RE))),
        fuel=_fuel_hint(raw),
    )


def compare_vehicle_identity_hints(
    lot_text: str | None, candidate_text: str | None
) -> dict[str, dict[str, object]]:
    lot = extract_vehicle_identity_hints(lot_text).as_dict()
    candidate = extract_vehicle_identity_hints(candidate_text).as_dict()
    out: dict[str, dict[str, object]] = {}
    for key in ("engine_cc", "transmission", "drivetrain", "fuel"):
        lot_value = lot[key]
        candidate_value = candidate[key]
        if lot_value is None:
            status = "LOT_UNKNOWN"
        elif candidate_value is None:
            status = "CANDIDATE_UNKNOWN"
        elif lot_value == candidate_value:
            status = "CONSISTENT"
        else:
            status = "DIFFERS"
        out[key] = {
            "lot": lot_value,
            "candidate": candidate_value,
            "status": status,
        }
    return out

File: supabase/functions/superbid-fasecolda-candidate-cockpit/identity_hints.ts
export const IDENTITY_HINT_GUARDRAIL = "AUTOMATED_IDENTITY_HINT_NOT_HUMAN_EVIDENCE_OR_MATCH";
export const ENGINE_CC_NOMINAL_TOLERANCE = 50;

export type IdentityHints = {
  engineCc: number | null;
  transmission: "MANUAL" | "AUTOMATIC" | null;
  drivetrain: "4X4_AWD" | "4X2_2WD" | null;
  fuel: "GASOLINE" | "DIESEL" | "HYBRID" | "ELECTRIC" | "CNG" | null;
};

export type HintComparisonStatus = "CONSISTENT" | "NOMINAL_COMPATIBLE" | "DIFFERS" | "LOT_UNKNOWN" | "CANDIDATE_UNKNOWN";
export type HintComparison = {
  lot: string | number | null;
  candidate: string | number | null;
  status: HintComparisonStatus;
};

const ENGINE_CC_RE = /\b(\d{3,5})\s*CC\b/gi;
const MANUAL_TRANSMISSION_RE = /\b(?:MT|MANUAL|MEC[ÁA]NIC[AO])\b/i;
const AUTOMATIC_TRANSMISSION_RE = /\b(?:AT|TP|CVT|DCT|DSG|AUT|AUTOM[ÁA]TIC[AO])\b/i;
const DRIVE_4X4_RE = /\b(?:4\s*[Xx]\s*4|4WD|AWD)\b/i;
const DRIVE_4X2_RE = /\b(?:4\s*[Xx]\s*2|2WD)\b/i;
const HYBRID_RE = /\b(?:H[IÍ]BRID[AO]|HYBRID|HEV|PHEV)\b/i;
const ELECTRIC_RE = /\b(?:EL[EÉ]CTRIC[AO]|ELECTRIC|EV|BEV)\b/i;
const DIESEL_RE = /\b(?:DI[EÉ]SEL|DIESEL)\b/i;
const CNG_RE = /\b(?:GNV|GNC|CNG|GAS\s+NATURAL)\b/i;
const GASOLINE_RE = /\b(?:GASOLINA|GASOLINE|PETROL)\b/i;

function uniqueEngineCc(text: string): number | null {
  const values = new Set<number>();
  for (const match of text.matchAll(ENGINE_CC_RE)) values.add(Number(match[1]));
  return values.size === 1 ? [...values][0] : null;
}

function exclusiveFlag<T extends string>(text: string, rules: Array<[T, RegExp]>): T | null {
  const values = new Set<T>();
  for (const [label, pattern] of rules) if (pattern.test(text)) values.add(label);
  return values.size === 1 ? [...values][0] : null;
}

function fuelHint(text: string): IdentityHints["fuel"] {
  if (HYBRID_RE.test(text)) return "HYBRID";
  if (ELECTRIC_RE.test(text)) return "ELECTRIC";
  return exclusiveFlag(text, [
    ["DIESEL", DIESEL_RE],
    ["CNG", CNG_RE],
    ["GASOLINE", GASOLINE_RE],
  ]);
}

export function extractVehicleIdentityHints(value: unknown): IdentityHints {
  const text = String(value ?? "");
  return {
    engineCc: uniqueEngineCc(text),
    transmission: exclusiveFlag(text, [
      ["MANUAL", MANUAL_TRANSMISSION_RE],
      ["AUTOMATIC", AUTOMATIC_TRANSMISSION_RE],
    ]),
    drivetrain: exclusiveFlag(text, [
      ["4X4_AWD", DRIVE_4X4_RE],
      ["4X2_2WD", DRIVE_4X2_RE],
    ]),
    fuel: fuelHint(text),
  };
}

function compareOne(lot: string | number | null, candidate: string | number | null): HintComparison {
  if (lot == null) return { lot, candidate, status: "LOT_UNKNOWN" };
  if (candidate == null) return { lot, candidate, status: "CANDIDATE_UNKNOWN" };
  return { lot, candidate, status: lot === candidate ? "CONSISTENT" : "DIFFERS" };
}

function compareEngineCc(lot: number | null, candidate: number | null): HintComparison {
  if (lot == null) return { lot, candidate, status: "LOT_UNKNOWN" };
  if (candidate == null) return { lot, candidate, status: "CANDIDATE_UNKNOWN" };
  if (lot === candidate) return { lot, candidate, status: "CONSISTENT" };
  if (Math.abs(lot - candidate) <= ENGINE_CC_NOMINAL_TOLERANCE) {
    return { lot, candidate, status: "NOMINAL_COMPATIBLE" };
  }
  return { lot, candidate, status: "DIFFERS" };
}

export function compareVehicleIdentityHints(lotText: unknown, candidateText: unknown) {
  const lot = extractVehicleIdentityHints(lotText);
  const candidate = extractVehicleIdentityHints(candidateText);
  return {
    engine_cc: compareEngineCc(lot.engineCc, candidate.engineCc),
    transmission: compareOne(lot.transmission, candidate.transmission),
    drivetrain: compareOne(lot.drivetrain, candidate.drivetrain),
    fuel: compareOne(lot.fuel, candidate.fuel),
  };
}
